Re-implement common interface bindings

PerezHz opened the pull request:
This PR proposes to re-implement the common interface bindings to the `DifferentialEquations.jl` ecosystem essentially by adding `TaylorMethod` as an ODE solver method of `OrdinaryDiffEq.jl` in the sense `TaylorAlgorithm <: OrdinaryDiffEqAdaptiveAlgorithm`. So the proposed approach is to extend `initialize!` and `perform_step!` methods in `OrdinaryDiffEq.jl`, while handling step size control via corresponding method extensions of `stepsize_controller!` and `step_accept_controller!`. By doing this, calling `solve` yields the same results as a direct call to `taylorinteg` with comparable performance, while allowing the use of a larger set of keywords with `TaylorMethod`, including `saveat`, `tstops` and even `callback` (currently, discrete callbacks seem to work okay while continuous callbacks do not).

On the `TaylorIntegration.jl` side, I extended `jetcoeffs!` and `stepsize`, as well as `TaylorSeries.evaluate!` to allow use of high-dimension arrays. Following this strategy, current tests in `master` pass, and also added some more tests. In particular, this PR might help to find a way out of #103 and #109.

There are some issues remaining: besides the continuous callbacks (which I'm currently trying to sort out) I'm not completely sure how to handle the FSAL stuff, but since it is used in 3rd-order Hermite interpolation in `OrdinaryDiffEq`, I added FSAL-related updates in `initialize!` and `perform_step!`, although I'm not sure if I did it right. Also, I left most of the keywords in the warning list `warnkeywords`, although a better approach might now be the inverse: to ditch the keyword warning list altogether (or at least most of it?), and handle any issues that will arise in the future.

Surely there are lots of things that can be improved, so feedback is appreciated!

cc: @ChrisRackauckas @lbenet

PerezHz commented:
Updates: I found a way, thanks to @SebastianM-C's work in #108, to finally handle `DynamicalODEProblem`s with TaylorIntegration.jl. A thing to note is that use of `@taylorize` for speedup is currently not allowed for DynamicalODEProblem. Anyway, with the latest changes this PR fixes #109. We can always take care of performance for DynamicalODEProblem, via use of `@taylorize`, in a future PR.

lbenet commented:
Thanks for the great effort! 

Please ping me, and perhaps @SebastianM-C too, whenever you think this is ready for review.

PerezHz commented:
> Please ping me, and perhaps @SebastianM-C too, whenever you think this is ready for review.

The latest update is that I think I found a way to make continuous and vector callbacks work (discrete callbacks were already working), which I hope will help users of TaylorIntegration.jl through the common interface.

I'm still not 100% percent sure about the FSAL thing, but things appear to be working well, so... 🤷‍♂️. The keyword warning list perhaps needs to be updated in some way as well, but here's where some feedback might help to know what's the best way to proceed going forward. I think we should add support, respectively, for in-place format in scalar problems and out-of-place format in array problems, but we might tackle that in another PR. Finally, in the tests some funny warnings started to appear:
```
WARNING: Method definition vec(Number) in module FiniteDiff at
C:\Users\appveyor\.julia\packages\FiniteDiff\EBys0\src\jacobians.jl:128 
overwritten in module DiffEqDiffTools at
C:\Users\appveyor\.julia\packages\DiffEqDiffTools\uD0fb\src\jacobians.jl:114.
```
but that seems to be related to FiniteDiff and DiffEqDiffTools, since we're not overwriting any `vec` methods here.

Otherwise, this PR is ready for review!

cc: @lbenet @SebastianM-C

SebastianM-C commented on src/common.jl:
```diff
@@ -1,194 +1,283 @@
+    sizeu = size(prob.u0)
+    # This if block handles DynamicalODEProblems
+    # credit to @SebastianM-C for coming up with this solution
+    if prob.f isa DynamicalODEFunction
```
From what I understand, this special casing is needed because the `jetcoeffs!` function calls the user function with `xaux` which is an `Array`. I was wondering if it would be possible to just use the `DiffEqFunction` interface (i.e. call `prob.f` with `f(du,u,p,t)` and `u` with the same type as `prob.u0`), but I'm not sure I understand how the `jetcoeffs!` code works and if the above suggestion would make sense.

PerezHz replied:
Thank you for your review and your suggestions! You're right, it'd be nice if we didn't have to convert everything to an `ODEProblem`... after reading your comment here I gave it a shot, and I think just got to overcome this conversion, at least for the in-place case (I should push that commit in a couple of minutes). The out-of-place version won't work as of now, since `TaylorIntegration.jl` currently does not support oop for `prob.u0 <: AbstractArray`. But once we add that, perhaps in another PR, we'll be able to get rid of most of the `if`s and `else`s, while avoiding the conversion from `DynamicalODEProblem` to `ODEProblem`.

SebastianM-C reviewed (commented):
Thanks for adding support for `DynamicalODEProblems`. I was thinking if a more general approach where the conversion to `ODEProblem` is no longer required would be possible.

PerezHz commented:
Just pushed a commit which fixed an error in the casing for out-of-place DynamicalODEProblem (the returned value from f wasn't being passed correctly) and added some tests.

PerezHz commented:
@SebastianM-C I was thinking... is there a way to convert an out-of-place `DynamicalODEProblem` to an in-place `DynamicalODEProblem`? Since we now can handle in-place `DynamicalODEProblem` if somehow we were able to do that conversion, that way we could avoid the conversion to `ODEProblem` for out-of-place `DynamicalODEProblem`...

PerezHz commented:
I was trying stuff like
```julia
if prob.f isa DynamicalODEFunction && !isinplace
    f1! = (dv, v, u, p, t) -> (dv .= prob.f.f1(v, u, p, t); 0)
    f2! = (du, v, u, p, t) -> (du .= prob.f.f2(v, u, p, t); 0)
    prob.f = DynamicalODEFunction{true}(f1!, f2!)
    # ...
```
which is similar to what we do for out-of-place `ODEProblem`, but then I get an error:
```
ERROR: MethodError: Cannot `convert` an object of type
  DynamicalODEFunction{true,ODEFunction{true,typeof(iip_ṗ),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},ODEFunction{true,typeof(iip_q̇),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}} to an object of type
  DynamicalODEFunction{false,ODEFunction{false,typeof(oop_ṗ),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},ODEFunction{false,typeof(oop_q̇),LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}},LinearAlgebra.UniformScaling{Bool{}},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{},Nothing{}}
Closest candidates are:
  convert(::Type{T}, ::T) where T at essentials.jl:171
  DynamicalODEFunction{false,ODEFunction{false,typeof(oop_ṗ),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},ODEFunction{false,typeof(oop_q̇),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing}(::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any) where {iip, F1, F2, TMM, Ta, Tt, TJ, JVP, VJP, JP, SP, TW, TWt, TPJ, S, TCV} at /Users/Jorge/.julia/packages/DiffEqBase/3iigH/src/diffeqfunction.jl:64
Stacktrace:
 [1] setproperty!(::ODEProblem{RecursiveArrayTools.ArrayPartition{Float64,Tuple{SArray{Tuple{2},Float64,1,2},SArray{Tuple{2},Float64,1,2}}},Tuple{Float64,Float64},false,DiffEqBase.NullParameters,DynamicalODEFunction{false,ODEFunction{false,typeof(oop_ṗ),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},ODEFunction{false,typeof(oop_q̇),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem}, ::Symbol, ::Function) at ./Base.jl:34
 [2] top-level scope at REPL[35]:1
```

In the tests, if I try

```julia
oop_prob.f = iip_prob.f
```
the same error shows up.

SebastianM-C commented on test/common.jl:
```diff
@@ -235,11 +241,11 @@ using StaticArrays
         function oop_ṗ(p, q, params, t)
             dp1 = -q[1] * (1 + 2q[2])
             dp2 = -q[2] - (q[1]^2 - q[2]^2)
-            @SVector [dp1, dp2]
```
Why did you remove `SVector` initial conditions?

PerezHz replied:
Thanks for checking this! I removed them temporarily because it doesn't work with latest commit 😅 since `setindex!` is not defined for `SArray{S,Taylor1}`, I'm trying to fix that right now, just pushed this to see how travis reacted

PerezHz replied:
I'll add them again as soon as I fix this

PerezHz replied:
They are back now... not the prettiest solution, but it was the best compromise until we properly support oop problems in TaylorIntegration.jl

SebastianM-C commented on src/common.jl:
```diff
@@ -186,40 +186,28 @@ function DiffEqBase.solve(
-        if !isinplace && typeof(prob.u0) <: AbstractArray
-            f! = (du, u, p, t) -> (du .= f(u, p, t); 0)
+    f = prob.f
+    if !isinplace && typeof(prob.u0) <: AbstractArray
```
Is `typeof(prob.u0) <: AbstractArray` necessary?
I think that the only case that we have `prob.u0` not be an array is when it is a scalar differential equation, but in that case it can only be out-of-place, so `!isinplace` is already true.

PerezHz replied:
Thank you for your suggestion. If we think an oop scalar problem, and if we removed that `typeof(prob.u0) <: AbstractArray`, then in that case the code would try to transform it into a in-place scalar problem, but currently TaylorIntegration does not support in-place scalar problems; that's why I left it there. Tests actually break if I remove that part

SebastianM-C reviewed (commented):
This is a much better solution as it preserves the problem type and avoids converting the user input to an `Array`. This should improve both performance and generality/composability.

SebastianM-C commented on src/common.jl:
```diff
@@ -1,194 +1,258 @@
+    f = prob.f
+    if !isinplace && typeof(prob.u0) <: AbstractArray
+        if prob.f isa DynamicalODEFunction
+            f1! = (dv, v, u, p, t) -> (dv .= prob.f.f1(v, u, p, t); 0)
```
Since we are in the out-of-place case, shouldn't this _not_ use broadcasting to modify the elements of `dv` in-place? 
```suggestion
            f1! = (dv, v, u, p, t) -> (dv = prob.f.f1(v, u, p, t); 0)
```

SebastianM-C commented on src/common.jl:
```diff
@@ -1,194 +1,258 @@
+    f = prob.f
+    if !isinplace && typeof(prob.u0) <: AbstractArray
+        if prob.f isa DynamicalODEFunction
+            f1! = (dv, v, u, p, t) -> (dv .= prob.f.f1(v, u, p, t); 0)
+            f2! = (du, v, u, p, t) -> (du .= prob.f.f2(v, u, p, t); 0)
```
Since we are in the out-of-place case, shouldn't this _not_ use broadcasting to modify the elements of `dv` in-place?
```suggestion
            f1! = (dv, v, u, p, t) -> (dv = prob.f.f1(v, u, p, t); 0)
            f2! = (du, v, u, p, t) -> (du = prob.f.f2(v, u, p, t); 0)
```

SebastianM-C commented on src/common.jl:
```diff
@@ -1,194 +1,258 @@
+    f = prob.f
+    if !isinplace && typeof(prob.u0) <: AbstractArray
+        if prob.f isa DynamicalODEFunction
+            f1! = (dv, v, u, p, t) -> (dv .= prob.f.f1(v, u, p, t); 0)
```
This is what causes test to fail in the OOP case with immutable types.

PerezHz replied:
Good catch, thanks! I'm proposing, as a temporary solution for this specific case, to convert the initial condition `prob.u0` from an immutable to a mutable array type (see below)

SebastianM-C reviewed (commented):
The translation of out-of-place problems to in-place ones is fundamentally incompatible with immutable types. From what I understand from the DiffEq code, the OOP code path is made specifically for immutable types.

PerezHz commented:
Ok, so I think I found a compromise: I found a way to make things to work with `SVectors` when dealing with oop `DynamicalODEProblem` by transforming specifically for this case the initial condition `prob.u0` from an immutable array type to a mutable one (it's highly non-elegant, but does the job). I fully agree that this does not solve the incompatibility of the oop->iip translation with immutable arrays, but solving that deeper issue requires some changes under the hood of TaylorIntegration.jl itself, which I believe we should think through and tackle in another PR.

lbenet commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ examples/.ipynb_checkpoints
 examples/JuliaCon2017/.ipynb_checkpoints/
 docs/build/
 docs/site/
+.vscode/
```
This may need  a new line (it's the end of the file); this is noted by github.

PerezHz replied:
Fixed, thanks!

lbenet commented on src/common.jl:
```diff
@@ -1,194 +1,267 @@
-     :qmax, :qmin, :qsteady_min, :qsteady_max, :qoldinit, :failfactor,
-     :maxiters, :isoutofdomain, :unstable_check,
-     :calck, :progress, :timeseries_steps, :tstops, :dense)
+import OrdinaryDiffEq: OrdinaryDiffEqAdaptiveAlgorithm,
+OrdinaryDiffEqConstantCache, OrdinaryDiffEqMutableCache,
+alg_order, alg_cache, initialize!, perform_step!, @unpack,
+@cache, stepsize_controller!, step_accept_controller!
+
+# TODO: check which keywords work fine
+const warnkeywords = (:save_idxs, :d_discontinuities, :unstable_check, :save_everystep,
+:save_end, :initialize_save, :adaptive, :dt, :reltol, :dtmax,
+:dtmin, :force_dtmin, :internalnorm, :gamma, :beta1, :beta2,
+:qmax, :qmin, :qsteady_min, :qsteady_max, :qoldinit, :failfactor,
+:isoutofdomain, :unstable_check,
+:calck, :progress, :timeseries_steps, :dense)
```
Can you include some indentation; makes nicer reading the code.

Can you open an issue to  have present the TO DO indicated here?

PerezHz replied:
Thanks, added indentation and also opened the issue corresponding to this TODO

lbenet commented on src/common.jl:
```diff
@@ -1,194 +1,267 @@
 struct TaylorMethod <: TaylorAlgorithm
     order::Int
 end
+struct _TaylorMethod <: TaylorAlgorithm
+    order::Int
+    parse_eqs::Bool
+end
+
+_TaylorMethod(order; parse_eqs=true) = _TaylorMethod(order, parse_eqs) # set `parse_eqs` to `true` by default
```
Naive question: why do we need two (essentially identical) structs, instead  of only one? Perhaps by adjusting some constructors, everything works with only one...

PerezHz replied:
Good catch! Yes, it was redundant to have those two structures, so I removed one of them

lbenet commented on src/common.jl:
```diff
@@ -1,194 +1,267 @@
+# interpolation may be helpful (since Taylor interpolation is memory-intensive),
+# so let's set isfsal to true for now, which is the default, so no method
+# definition needed here, so we in principle should support FSAL
+# isfsal(::TaylorMethod) = true # see discussion above
```
Yet, this line is commented... Is it needed? We can certainly have it there commented, but then clarifying a bit would be very enlightening.

PerezHz replied:
It was more of a personal reminder when trying to figure the internals of DiffEqBase and OrdinaryDiffEq; I've now removed these commented lines

lbenet commented on src/common.jl:
```diff
@@ -1,194 +1,267 @@
+end
 
-    tspan = prob.tspan
+# @show macroexpand(@__MODULE__, :(@cache struct TaylorMethodCache{uType,rateType,tTType,uTType} <: OrdinaryDiffEqMutableCache
+#   u::uType
+#   uprev::uType
+#   tmp::uType
+#   k::rateType
+#   fsalfirst::rateType
+#   tT::tTType
+#   uT::uTType
+#   duT::uTType
+#   uauxT::uTType
+#   parse_eqs::Bool
+# end) )
```
Do we need this? Seems it is a left-over of something you were testing...

PerezHz replied:
😅 sorry, thanks! Yes, this was a leftover, it's now gone

lbenet commented on src/common.jl:
```diff
@@ -1,194 +1,267 @@
+# and vector callbacks with TaylorIntegration.jl via the common interface
+function DiffEqBase.addsteps!(k,t,uprev,u,dt,f,p,cache::TaylorMethodCache,always_calc_begin = false,allow_calc_end = true,force_calc_end = false)
+    if length(k)<2 || always_calc_begin
+        if typeof(cache) <: OrdinaryDiffEqMutableCache
+        rtmp = similar(u, eltype(eltype(k)))
+        f(rtmp,uprev,p,t)
+        copyat_or_push!(k,1,rtmp)
+        f(rtmp,u,p,t+dt)
+        copyat_or_push!(k,2,rtmp)
+        else
+        copyat_or_push!(k,1,f(uprev,p,t))
+        copyat_or_push!(k,2,f(u,p,t+dt))
```
Can you include some indentation here?

PerezHz replied:
Thanks, done!

lbenet commented on src/common.jl:
```diff
@@ -1,194 +1,267 @@
-      push!(d_discontinuities_internal, tdir * t)
-    end
+# This function was modified from TaylorSeries.jl; MIT-licensed
+# evaluate! overload to handle AbstractArray
+import TaylorSeries: evaluate!
+function evaluate!(x::AbstractArray{Taylor1{T},N}, δt::S,
+        x0::AbstractArray{T,N}) where {T<:Number, S<:Number, N}
 
-    tstops_internal, saveat_internal, d_discontinuities_internal
+    # @assert length(x) == length(x0)
+    @inbounds for i in eachindex(x, x0)
+        x0[i] = evaluate( x[i], δt )
+    end
+    nothing
 end
```
Why not  adding it directly to TaylorSeries?

PerezHz replied:
This is the only point which I haven't addressed yet; I will open the corresponding PR in TaylorSeries, and it will perhaps require a new TaylorSeries release

PerezHz replied:
Update: just opened the corresponding PR. See: https://github.com/JuliaDiff/TaylorSeries.jl/pull/252

lbenet commented on test/common.jl:
```diff
@@ -1,16 +1,24 @@
-using TaylorIntegration, Test, DiffEqBase
+using Test, DiffEqBase, TaylorIntegration
```
Does the order matter here? If so, i think it is worth adding  a couple of sentences to the docs.

PerezHz replied:
The order makes no difference, I just forgot to leave that line untouched. I've now gone back to the original order, so that this line doesn't show up when doing diffs in the future

lbenet commented on src/common.jl:
```diff
@@ -1,194 +1,267 @@
-      end
+    # DiffEqBase.solve(prob, _alg, args...; kwargs...)
+    integrator = DiffEqBase.__init(_prob, _alg, args...; kwargs...)
+    integrator.dt = stepsize(integrator.cache.uT, integrator.opts.abstol) # override handle_dt! setting of initial dt
```
Previously, there was a default value set for `abstol`; I cannot see it now. So the question is if the user must now to specify it, or is it set somewhere else?

PerezHz replied:
Good catch! If the user doesn't specify `abstol` it is set automatically by `DiffEqBase.__init` to `abstol=1e-6`

lbenet reviewed (approved):
Thanks for the tremendous work  that there is in this PR! I have left mostly trivial remarks  to clarify a couple of points. 

LGTM and I am in favor of merging this, though maybe we may wait for @SebastianM-C's opinion.

PerezHz commented:
> Thanks for the tremendous work that there is in this PR! I have left mostly trivial remarks to clarify a couple of points.

Thanks @lbenet for your review! I addressed all your comments except one, which will have to wait a bit while we add the new `evaluate!`method to TaylorSeries. In the meantime, please let me know if there's something that you think can be further improved.
 
> LGTM and I am in favor of merging this, though maybe we may wait for @SebastianM-C's opinion.

I agree we should wait to hear Sebastian's opinion.

lbenet commented:
Tests are passing, except in nightly; I am just including the (first) problem spotted by travis [here](https://travis-ci.org/github/PerezHz/TaylorIntegration.jl/jobs/747480423#L598) to address it later.

lbenet commented:
Shall I go ahead and merge?

PerezHz commented:
> Shall I go ahead and merge?

Sorry for the late reply, sure, I agree that we should merge!

lbenet commented:
Merged! Thanks a lot!
